projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs

```diff
@@ -126,36 +113,37 @@ async Task Loop()
 
             async Task LoopWithConcurrency(CancellationToken cancellationToken)
             {
-                while (_tokenSource.IsCancellationRequested == false)
+                try
                 {
-                    try
-                    {
-                        await _syncSource.Task.ConfigureAwait(false);
-                        _syncSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
-                    }
-                    catch (TaskCanceledException)
+                    while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
                     {
-                        // Swallowing the task cancellation exception for the semaphore in case we are stopping.
-                    }
-
-                    while (_actions.TryDequeue(out Action action))
-                    {
-                        // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
-                        if(!_limiter.Wait(0))
+                        while (_channel.Reader.TryRead(out Action action))
                         {
-                            await _limiter.WaitAsync(cancellationToken).ConfigureAwait(false);
-                        }
+                            // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
+                            if(!_limiter.Wait(0))
+                            {
+                                await _limiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+                            }
 
-                        _ = OffloadToWorkerThreadPool(action, _limiter);
+                            _ = OffloadToWorkerThreadPool(action, _limiter);
+                        }
                     }
                 }
+                catch (OperationCanceledException)
+                {
+                    // ignored
+                }
             }
 
             static async Task OffloadToWorkerThreadPool(Action action, SemaphoreSlim limiter)
             {
                 try
                 {
-                    await Task.Run(() => action());
+                    // like Task.Run but doesn't closure allocate
```

Comment:
🎉  I like =)

Reply:
Unfortunately, it still allocates delegate. IIRC it is well known issue of Roslyn. If you really do not want allocate delegate each call then define static delegate and use it here.

Reply:
@sungam3r thanks for chiming in. You are welcome to submit a PR that changes the comment (and links to the Roslyn issue) or even switched to use a static delegate if it's a safe thing to do in this context. Thank you.

Reply:
Comparison according to sharplab

```
    [Serializable]
    [CompilerGenerated]
    private sealed class <>c
    {
        public static readonly <>c <>9 = new <>c();

        public static Action<object> <>9__0_0;

        internal void <TaskFactory>b__0_0(object state)
        {
            ((Action)state)();
        }
    }

awaiter = Task.Factory.StartNew(<>c.<>9__0_0 ?? (<>c.<>9__0_0 = new Action<object>(<>c.<>9.<TaskFactory>b__0_0)), action, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default).GetAwaiter();
```

so the `Action` allocation happens once and is then cached for further execution in a static Action field.

vs `Task.Run(() => action())`

```
                    <>c__DisplayClass1_0 <>c__DisplayClass1_ = new <>c__DisplayClass1_0();
                    <>c__DisplayClass1_.action = action;
                    awaiter = Task.Run(new Action(<>c__DisplayClass1_.<TaskRun>b__0)).GetAwaiter();
```

display class creation for every run plus action allocation for every run

Reply:
I was wrong. Such lambda usage is OK. I re-read the Roslyn issue and remembered details.